File: 2021/java/day3/Day3.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class Day3 {

    public static void main(String[] args) {

        // Parse input
        List<String> binaryNumbers = read_input();

        // Extract the column bits
        List<List<Character>> columnBits = new ArrayList<List<Character>>();
        for (int i = 0; i < binaryNumbers.get(0).length(); i++) {
            columnBits.add(i, extract_column(binaryNumbers, i));
        }

        // Part 1
        String gamma = "";
        String epsilon = "";
        for (List<Character> column : columnBits) {
            int zeroes = count(column, '0');
            int ones = count(column, '1');
            String mc = zeroes >= ones ? "10" : "01";
            gamma += mc.charAt(0);
            epsilon += mc.charAt(1);
        }
        System.out.println("Part 1: " + (Integer.parseInt(gamma, 2) * Integer.parseInt(epsilon, 2)));

        // Part 2
        int ogr = calc_rating(binaryNumbers, "OGR");
        int cosr = calc_rating(binaryNumbers, "COSR");
        System.out.println("Part 2: " + (ogr * cosr));

    }

    public static int calc_rating(List<String> binaryNumbers, String typeOfRating) {
        int bit = 0;
        List<String> picked = binaryNumbers;
        while (picked.size() > 1) {
            List<Character> bits = extract_column(picked, bit);
            int zeroes = count(bits, '0');
            int ones = count(bits, '1');
            int mc = typeOfRating.equals("OGR") ? ones >= zeroes ? 1 : 0 : ones >= zeroes ? 0 : 1;
            int bit_ = bit;
            picked = picked.stream().filter(bin_num -> bin_num.charAt(bit_) == (char) (mc + '0'))
                    .collect(Collectors.toList());
            bit++;
        }
        return Integer.parseInt(picked.get(0), 2);
    }

    public static int count(List<Character> li, Character k) {
        int i = 0;
        for (Character bit : li) {
            if (bit.equals(k))
                i++;
        }
        return i;
    }

    public static List<Character> extract_column(List<String> li, int column) {
        List<Character> col = new ArrayList<Character>();
        for (String bin_num : li) {
            col.add(bin_num.charAt(column));
        }
        return col;
    }

    public static List<String> read_input() {
        try {
            List<String> instructions = new ArrayList<String>();
            File fileObj = new File("day3_input.txt");
            Scanner reader = new Scanner(fileObj);
            while (reader.hasNextLine()) {
                instructions.add(reader.nextLine().strip());
            }
            reader.close();
            return instructions;
        } catch (FileNotFoundException e) {
            System.out.println("An error occured.");
            e.printStackTrace();
            return Collections.emptyList();
        }
    }

}

File: 2021/python/day_3/day3.py
def run():

    # Utility methods
    def extract_column_bits(li, col):
        # Collect the bits from selected column
        return list(map(lambda bin_num: int(bin_num[col]), li))

    def most_common(li, comparator):
        zeros = li.count(0)
        ones = li.count(1)
        return comparator(ones, zeros)

    # Calculates the Oxygen generator- and CO2 scrubber ratings
    def calc_rating(comparator):
        picked = binary_numbers
        bit = 0
        while len(picked) > 1:
            bits = extract_column_bits(picked, bit)
            mc = most_common(bits, comparator)
            picked = list(filter(lambda num: num[bit] == mc, picked))
            bit += 1
        return int(picked[0], 2)

    # Parse input
    with open("day3_input.txt") as f:
        binary_numbers = f.read().split("\n")

    # Part 1
    gamma = ""
    epsilon = ""
    for column in [
        extract_column_bits(binary_numbers, i) for i in range(len(binary_numbers[0]))
    ]:
        mc = most_common(column, lambda a, b: ("1", "0") if a > b else ("0", "1"))
        gamma += mc[0]
        epsilon += mc[1]
    print(f"Part 1: {int(gamma, 2) * int(epsilon, 2)}")

    # Part 2
    ogr = calc_rating(lambda a, b: "1" if a >= b else "0")
    cosr = calc_rating(lambda a, b: "0" if a >= b else "1")
    print(f"Part 2: {ogr * cosr}")


if __name__ == "__main__":
    run()
